src/transform/src/attribute/mod.rs

```diff
@@ -98,98 +98,109 @@ impl<A: Attribute> Env<A> {
 impl<A: Attribute> Env<A> {
     /// Schedules environment maintenance tasks.
     ///
-    /// Should be called from a `Visitor<MirRelationExpr>::pre_visit` implementaion.
+    /// Should be called from a `Visitor<MirRelationExpr>::pre_visit` implementation.
     pub fn schedule_tasks(&mut self, expr: &MirRelationExpr) {
         match expr {
             MirRelationExpr::Let { id, .. } => {
-                // Add an Extend task to be handled in the post_visit the node children.
+                // Add a Remove and an Extend task to be handled in the post_visit of the node's
+                // children. The Extend will be found after completing the `value`, and the `Remove`
+                // will be found after completing the body.
+                self.env_tasks.push(EnvTask::Remove(vec![id.clone()]));
```

Comment:
These code changes break the original assumptions of that code:

1. At every `pre_visit` call, we push install exactly one `EnvTask` that represents of the state machine associated with the current node.
2. At every `post_visit` call, we will (a) pop the task associated with the just visited node, and (b) advance the state machine of our parent by updating the parent task.

I would like to propose an alternative version of step (3) that maintains this invariant (I will also try to make those more explicit in the code comments). I will push a fixup commit to this PR and then we can discuss which one is easier to maintain.